lep/DRAFT/012_Pro_version_personal_server.md

```diff
@@ -0,0 +1,142 @@
+
+This request needs to carry also some data to identify the Lantern instance (lantern id and IP).  For this, we can use Stripe request [metadata](https://stripe.com/docs/api#metadata).  Ideally, this will also provide geolocation and/or IP information.  The geolocation data will be useful for deciding in which DC to launch the proxy.
+
+
+#### Payments Server receives confirmation
+
+Once all validation has be done at Stripe's side, the UI Javascript side will receive a one-time token that will then be sent to the Lantern's Payments server so it can proceed with charging.
+
+**Important Note: Ideally, the user shouldn't be required to contact Lantern's Payments server.  Unfortunately there is currently no way to perform charges from the client, nor add a webhook at the validation stage.**
+
+At this point, the charge has been done and we can proceed with launching a server.  Stripe's charging API is *synchronous*, so if it succeeds we can act immediately, whithout requiring any callback to check status. As soon as the `charge` call returns, we are good to go **(step 4)**.  This will be done by appending a new Lantern Pro Server request in Redis DB.
+
+**Notes on optional features**
+
+Should card details be saved? We don't have users at the moment, and since we are working with tokens, it is probably best to keep it simple without users.  Additionally, it seems that Alipay users are more confortable with one-time payments than subscriptions, even if for larger periods.  Although they are separate things, the fact that users don't like subscriptions suggests that is better to just provide a very simple UI and no information saving or accounts.
+
+
+### 3. Providing a private Pro Server to the client
+
+#### Redis Pro Servers queue and management
+
+If the payment was successful, the charge has been registered and we can safely launch the server and provide the server and the client with the necessary information for secure proxying.  Thus, the Lantern Payments Server needs to append a new server request in a queue in Redis **(step 5 - a)**.  This will allow *Cloudmaster* to pick up this job and execute the requests in the best datacenter for the Lantern client.  The Redis DB is effectivelly serving as a message queue between the Payments Server and the Cloudmaster, just as what happens with the Config Server and Cloudmaster for the free service.
+
+Cloudmaster will need to run new tasks.  These are inevitable:
+
+1. Watch for new Pro Servers requests, and launch as necessary.  This is currently done for free LCSs, so it could extend the functionality or be based on it.
+
+1. Check private proxy remaining time of service, and shut down those that surpassed a certain time threshold.  Similar to what retiring servers is doing at the moment, but triggered by a time event.
```

Comment:
Yes, that is a good point. There are some things like this that I left undefined because they feel like business decisions and they don't imply any change in the architecture. I'll add this one because it is almost mandatory anyhow.
Thanks!